### ISSUE TITLE
commit c9905341ea34 ("Pass FUSE_PARALLEL_DIROPS to kernel (#861)") breaks backward comatibility

### ISSUE THREAD

**Issue by szmi** [libfuse/libfuse#879]
FUSE_CAP_PARALLEL_DIROPS is set by default.  Prior to this commit this had no effect since the respective flag wasn't set in the INIT reply.  After this change the flag is set by default if the kernel supports this feature, which means that any filesystem implementation that doesn't explicitly disable this feature will receive parallelized directory operations.  This might be okay in some cases, but if the filesystem is relying on serialization between these ops, than this change may result in breakage, possibly a crash.

This could be fixed by not setting this feature by default, so that it's only enabled if the filesystem explicitly requests it.  This was the intention of adding this flag on the kernel  interface.

**Comment by Nikratio** [comment]
You're right, I didn't think about that when merging.

On the other hand, the documentation has always clearly said:  "This feature is enabled by default when supported by the kernel". 

So I'm not sure if the fact that this has not been working correctly for a while is reason enough to deviate from the documented behaviour instead of just fixing the bug...

**Comment by szmi** [comment]
I'd opt for being cautious here.  The original change would result in non-obvious and hard to trigger bugs.  One option is to tie this to a minor version update, which means that at least a recompile is needed.  But even that is risky IMO.

Letting filesystems set this flag explicitly for more performance should be the safest bet.

**Comment by ashleypittman** [comment]
This request has come from some work I've been doing - we always check the FUSE_PARALLEL_DIROPS flag thinking it was enabled, when we saw the recent commit to libfuse to actually set it we did some testing and found that our code was asserting with the kernel flag enabled, but once we'd resolved that issue we do see significant improvement with this feature so we'd like to enable it - but only with new releases.  To be explicit - our code has never set this flag but we do manually read all the flags and as this was default on we log that and had expected that behaviour - even though in practice our code failed with it, we simply weren't testing what we thought we were.

I don't see an obviously best route forward here, even changing FUSE_CAP_PARALLEL_DIROPS from default on to default off doesn't guarantee safety in all cases does it?  There may be code out there that sets it in the wants flag but does not correctly work with it due to bugs.
